rotkehlchen/db/dbhandler.py

```diff
@@ -2233,6 +2233,99 @@ def get_netvalue_data(
                 data.append(total)
         return times_int, data
 
+    @staticmethod
+    def _infer_zero_timed_balances(
+            cursor: 'DBCursor',
+            balances: list[SingleDBAssetBalance],
+            from_ts: Optional[Timestamp] = None,
+            to_ts: Optional[Timestamp] = None,
+    ) -> list[SingleDBAssetBalance]:
+        """
+        Given a list of asset specific timed balances, infers the missing zero timed balances
+        for the asset. We add 0 balances on the start and end of a period of 0 balances.
+
+        Example
+        We have the following timed balances for ETH (value, time):
+        (1, 1), (1, 2), (2, 3), (5, 7), (5, 12)
+        The timestamps of all timed balances in the DB are:
+        (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)
+        So we need to infer the following zero timed balances:
+        (0, 4), (0, 6), (0, 8), (0, 11)
+        """
+        if len(balances) == 0:
+            return []
+
+        if from_ts is None:
+            from_ts = Timestamp(0)
+        if to_ts is None:
+            to_ts = ts_now()
+
+        # else
+        cursor.execute(
+            'SELECT COUNT(*) FROM timed_balances WHERE timestamp BETWEEN ? AND ?',
+            (from_ts, to_ts),
+        )
+        num_all_balances = cursor.fetchone()[0]
+
+        if len(balances) == num_all_balances:
+            return []
+
+        # else
+        asset_timestamps = [b.time for b in balances]
+        cursor.execute(
+            'SELECT timestamp, category FROM timed_balances WHERE timestamp BETWEEN ? AND ? '
+            'ORDER BY timestamp ASC',
+            (from_ts, to_ts),
+        )
+        all_timestamps, all_categories = zip(*cursor)
+        # dicts maintain insertion order in python 3.7+
+        timestamps_have_asset_balance = dict.fromkeys(all_timestamps, False)
+        timestamps_with_asset_balance = dict.fromkeys(asset_timestamps, True)
+        timestamps_have_asset_balance.update(timestamps_with_asset_balance)
+        prev_has_asset_balance = timestamps_have_asset_balance[all_timestamps[0]]  # the value of the first timestamp  # noqa: E501
+        prev_timestamp = all_timestamps[0]
+        inferred_balances: list[SingleDBAssetBalance] = []
+        is_zero_period_open = False
+        last_asset_category = all_categories[-1]  # just a placeholder value, no need to calculate the actual value here  # noqa: E501
+        for idx, (timestamp, has_asset_balance) in enumerate(timestamps_have_asset_balance.items()):  # noqa: E501
```

Comment:
This iterates over all of the balances returned from the previous sql query.

This can be a long iteration if the graph has been going on for 5 years, like some of our first users.

Why not limit this from the sqlite query itself?

And you seem to be getting the specific asset balances as an argument to this function called `balances`. I think it should be possible to do this entire logic in sqlite itself. Otherwise you are loading potentially a ton of data in memory here and iterating over all of them at each and every api endoint call. So each time the graph moves etc.

Reply:
As discussed extensively in discord both python and sqlite approach turned out to be pretty similar. @dimyG implemented it and run the appropriate timing tests: https://discord.com/channels/657906918408585217/1080783409640648765/1117864559185768548

So we ended up going only with python approach and as a configurable setting that is off by default due to slowing down graphs quite a bit.